### DISCUSSION TITLE
Does flit only include a single module matching the project name?

### DISCUSSION THREAD

**Discussion by diazona** [takluyver/flit#655] [Q&A] [answered]
Hello! I'm trying to help improve the packaging workflow of a project that uses flit, called [offpunk](https://git.sr.ht/~lioploum/offpunk). The project's structure is, shall we say, a bit _minimalist_: it puts all its `.py` files directly at the top level of the Git repository, with no `src` directory, no `__init__.py`, no packages of any sort. When I build a wheel from this project, the only Python file that gets included in the wheel is the one that matches the project's name, `offpunk.py`. So naturally I went looking for a way to change that, but I'm not finding one. In fact, I get the impression from the documentation that flit is specifically meant not to handle that case, although I haven't found it to be clearly stated anywhere. So my question is:

- Is it accurate that, when building a wheel, flit is only capable of including a single module or package that matches the project name?
- And if so, was that an intentional design choice? (more for my curiosity than anything else)

I figured I'd ask here so I can make sure I'm passing on the right information to the person I've been discussing with who works on the project.

Thanks for the work you do!

**Comment by takluyver** [comment, marked as answer]
Yes, that's right. It is a design decision - I think if you've got more
than 1 module, putting them into a package is usually a good idea (maybe
not always, but there are other packaging tools people can use if Flit
isn't suitable).

The import name matches the project name by default, but you can specify a
different name if you want. E.g. my project pynsist uses nsist as the
import name. But it still has to be one module/package.

This sentence in the docs is meant to say this: "Flit packages a single
importable module or package at a time..." But it's a tricky thing to say
precisely and clearly.

**Reply by diazona** [reply]
Makes sense. Thanks! I will take this back to the discussion, that they have to either put their files in a package or switch to a build system that supports multiple top-level files.

By the way, if it helps, I think there were three things that confused me about this:

- In the sentence you quoted, I think "at a time" subconsciously got me thinking that Flit could somehow package multiple importable modules or packages, but would do them separately, whatever that means.
- I guess I've somehow come to expect that the documentation for a build backend will have a pretty prominent section describing how it chooses which files to include, and I was thrown off when I didn't seem to find that.
- And there's also the fact that the sdist includes all the Python scripts but the wheel doesn't - which isn't Flit's fault, of course, but I did find it confusing.

If you'd be interested, I'd be happy to offer my thoughts on how to clarify this in the documentation as a PR.
